File: su/2026-04/20_boj_괄호추가하기.java
import java.io.BufferedReader;
import java.io.InputStreamReader;

// public class Main {
class Main {
    static int N;
    static char[] expression;
    static long answer = Long.MIN_VALUE;

    static long calc(long a, char op, long b) {
        if (op == '+') return a + b;
        if (op == '-') return a - b;
        return a * b;
    }

    static void dfs(int i, long result) {
        if (i >= N) {
            answer = Math.max(answer, result);
            return;
        }

        // 괄호 안 치는 경우
        long noBracket = calc(result, expression[i], expression[i + 1] - '0');
        dfs(i + 2, noBracket);

        // 괄호 치는 경우
        if (i + 3 < N) {
            long yesBracket = calc(expression[i + 1] - '0', expression[i + 2], expression[i + 3] - '0');
            dfs(i + 4, calc(result, expression[i], yesBracket));
        }
    }

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        N = Integer.parseInt(br.readLine());
        expression = br.readLine().toCharArray();

        dfs(1, expression[0] - '0');

        System.out.println(answer);
    }
}

File: su/2026-04/20_boj_괄호추가하기.py
'''
중첩된 괄호는 사용할 수 없다.
시간초과가 나려나 dfs bfs
연산자 우선순위는 동일하기 때문에, 수식을 계산할 때는 왼쪽에서 순서대로 계산한다.
'''

N = int(input())
expression = list(input().strip())
answer = -int(1e9)

def calc(a, op, b):  
    a = int(a)
    b = int(b)
    if op == '+':
        return a + b
    elif op == '-':
        return a - b
    else: 
        return a * b

def dfs(i, result):  
    global answer
    if i >= N:
        answer = max(answer, result)
        return
    
    # 괄호 안 치는 경우
    no_bracket = calc(result, expression[i], expression[i+1])  
    dfs(i + 2, no_bracket)

    # 괄호를 치는 경우
    if i + 3 < N:
        yes_bracket = calc(expression[i+1], expression[i+2], expression[i+3])
        dfs(i + 4, calc(result, expression[i], yes_bracket))
    
    
dfs(1, expression[0])
print(answer)